game/songs.cc

```diff
@@ -151,16 +157,19 @@ void Songs::filter_internal() {
 	RestoreSel restore(*this);
 	try {
 		SongVector filtered;
-		for (SongVector::const_iterator it = m_songs.begin(); it != m_songs.end(); ++it) {
-			Song& s = **it;
-			// All, Dance, Vocals, Duet, Guitar, Band
-			if (m_type == 1 && !s.hasDance()) continue;
-			if (m_type == 2 && !s.hasVocals()) continue;
-			if (m_type == 3 && !s.hasDuet()) continue;
-			if (m_type == 4 && !s.hasGuitars()) continue;
-			if (m_type == 5 && !s.hasDrums() && !s.hasKeyboard()) continue;
-			if (m_type == 6 && (!s.hasVocals() || !s.hasGuitars() || (!s.hasDrums() && !s.hasKeyboard()))) continue;
-			if (regex_search(s.strFull(), boost::regex(m_filter, boost::regex_constants::icase))) filtered.push_back(*it);
+		if (m_filter == std::string() && m_type == 0) filtered = m_songs;
+		else {
+			icu::UnicodeString filter = icu::UnicodeString::fromUTF8(m_filter);
+			std::copy_if (m_songs.begin(), m_songs.end(), std::back_inserter(filtered), [&](boost::shared_ptr<Song> it){
```

Comment:
this is a bit hard to read since a lot is going on in a single command! (lambda function)

Reply:
imho it's easier to read if you're used to lambda's. Within c# we use them all the time (LINQ), besides the compiler can more easy optimize it.